pkg/mimirpb/timeseries.go

```diff
@@ -71,14 +78,111 @@ type PreallocTimeseries struct {
 	// If it is set to a non-nil value then it must be returned to the yoloSlicePool on cleanup,
 	// if it is set to nil then it can be ignored because the backing byte slice came from somewhere else.
 	yoloSlice *[]byte
+
+	// Original data used for unmarshalling this PreallocTimeseries. When set, Marshal methods will return it
+	// instead of doing full marshalling again. This assumes that this instance hasn't changed.
+	unmarshalData []byte
+}
+
+// RemoveLabel removes the label labelName from this timeseries, if it exists.
+func (p *PreallocTimeseries) RemoveLabel(labelName string) {
+	for i := 0; i < len(p.Labels); i++ {
+		pair := p.Labels[i]
+		if pair.Name == labelName {
+			p.Labels = append(p.Labels[:i], p.Labels[i+1:]...)
+			p.clearUnmarshalData()
+			return
+		}
+	}
+}
+
+func (p *PreallocTimeseries) SetLabels(lbls labels.Labels) {
+	p.Labels = FromLabelsToLabelAdapters(lbls)
+
+	// We can't reuse raw unmarshalled data for the timeseries after setting new labels.
+	// (Maybe we could, if labels are exactly the same, but it's expensive to check.)
+	p.clearUnmarshalData()
+}
+
+// RemoveEmptyLabelValues remove labels with value=="" from this timeseries, updating the slice in-place.
+func (p *PreallocTimeseries) RemoveEmptyLabelValues() {
+	modified := false
+	for i := len(p.Labels) - 1; i >= 0; i-- {
+		if p.Labels[i].Value == "" {
+			p.Labels = append(p.Labels[:i], p.Labels[i+1:]...)
+			modified = true
+		}
+	}
+	if modified {
+		p.clearUnmarshalData()
+	}
+}
+
+// SortLabelsIfNeeded sorts labels if they were not sorted before.
+func (p *PreallocTimeseries) SortLabelsIfNeeded() {
+	// no need to run sort.Slice, if labels are already sorted, which is most of the time.
+	// we can avoid extra memory allocations (mostly interface-related) this way.
+	sorted := true
+	last := ""
+	for _, l := range p.Labels {
+		if last > l.Name {
+			sorted = false
+			break
+		}
+		last = l.Name
+	}
+
+	if sorted {
+		return
+	}
+
+	sort.Slice(p.Labels, func(i, j int) bool {
+		return p.Labels[i].Name < p.Labels[j].Name
+	})
+	p.clearUnmarshalData()
+}
+
+// clearUnmarshalData removes cached unmarshalled version of the message.
+func (p *PreallocTimeseries) clearUnmarshalData() {
+	p.unmarshalData = nil
 }
 
 // Unmarshal implements proto.Message.
 func (p *PreallocTimeseries) Unmarshal(dAtA []byte) error {
+	p.unmarshalData = dAtA
 	p.TimeSeries = TimeseriesFromPool()
 	return p.TimeSeries.Unmarshal(dAtA)
 }
 
+func (p *PreallocTimeseries) Size() int {
+	if p.unmarshalData != nil {
+		return len(p.unmarshalData)
+	}
+	return p.TimeSeries.Size()
+}
+
+func (p *PreallocTimeseries) Marshal() ([]byte, error) {
+	if p.unmarshalData != nil {
+		return p.unmarshalData, nil
+	}
+	return p.TimeSeries.Marshal()
+}
+
+func (p *PreallocTimeseries) MarshalTo(dAtA []byte) (int, error) {
+	if p.unmarshalData != nil {
+		return p.MarshalToSizedBuffer(dAtA)
```

Comment:
`MarshalToSizedBuffer()` expects the input slice to be sized, while `MarshalTo()` doesn't. Here we call `MarshalToSizedBuffer()` assuming it's sized but it may be not. I think we should resize it here, before calling `MarshalToSizedBuffer()`.

Reply:
I only called `p.MarshalToSizedBuffer` to reuse the same copying logic. I'll just copy that code instead.

Reply:
https://github.com/grafana/mimir/pull/5137/commits/dda9aae79170c77d823193a914cd821699e5186a